### PR TITLE
[GLUTEN-2152][CORE] Skip local sort if child output ordering satisfies topk

### PR DESCRIPTION
**Pull request by ulysses-you**
## What changes were proposed in this pull request?

(Fixes: \#2152)

## How was this patch tested?

Pass existed topk test

### CODE REVIEW

**Comment by rui-mo** [comment]
@lgbo-ustc Would you like to take a review?

**Comment by lgbo-ustc** [comment]
LGTM

**Comment by lgbo-ustc** [comment]
LGTM
BTW, is there any unit tests to cover this ?

**Comment by ulysses-you** [comment]
@lgbo-ustc due to it changes plan inside transformer, it's a bit hard to plan validation.  I have tested this locally:

```
val row = 10 * 1000
val df1 = spark.range(0, row, 1, 2).selectExpr("id % 3 as c1")
val df2 = spark.range(0, row, 1, 2).selectExpr("id % 3 as c2")
df1.join(df2, col("c1") === col("c2")).orderBy(col("c1")).limit(100).write.format("noop").mode("Overwrite").save()

// before
7.481942 s
// after 
9.447648 s
```

File: gluten-core/src/main/scala/io/glutenproject/execution/TakeOrderedAndProjectExecTransformer.scala [hunk 1]
```diff
@@ -68,32 +67,40 @@ case class TakeOrderedAndProjectExecTransformer(limit: Int,
     if (childRDDPartsNum == 0) {
       sparkContext.parallelize(Seq.empty, 1)
     } else {
+      val orderingSatisfies = SortOrder.orderingSatisfies(child.outputOrdering, sortOrder)
+      def withLocalSort(sparkPlan: SparkPlan): SparkPlan = {
+        if (orderingSatisfies) {
+          sparkPlan
+        } else {
+          SortExecTransformer(sortOrder, false, sparkPlan)
```
Comment (zhztheplayer) [inline gluten-core/src/main/scala/io/glutenproject/execution/TakeOrderedAndProjectExecTransformer.scala:75]: Just out of curiosity (lack of some backgrounds), does the code mean we are not using optimized algorithms for the top-k calculation here? I noticed that vanilla Spark calls Guava's top-k which is ideally O(n).

@lgbo-ustc

Comment (ulysses-you) [reply gluten-core/src/main/scala/io/glutenproject/execution/TakeOrderedAndProjectExecTransformer.scala:75]: It seems velox support topN. I'm not sure why we did not support transform TakeOrderedAndProjectExec to TopN before. This pr itself just removes a local sort based on previous code.

**Comment by liuneng1994** [comment]
The optimization on one partition also failed on tpcds q67, should we remove directly

**Comment by rui-mo** [comment]
> The optimization on one partition also failed on tpcds q67, should we remove directly

@lgbo-ustc @liuneng1994 Can we merge this PR? Is there any remaining issue on CH backend?

**Comment by zzcclp** [comment]
I will take a review ASAP, if ok I will merge this.
BTW, @ulysses-you please help to open tpcds q67 test by this way:
```
\\ in GlutenClickHouseTPCDSAbstractSuite.scala line 98:
remove "q67"
```

**Comment by zzcclp** [comment]
> I will take a review ASAP, if ok I will merge this. BTW, @ulysses-you please help to open tpcds q67 test by this way:
> 
> ```
> \\ in GlutenClickHouseTPCDSAbstractSuite.scala line 98:
> remove "q67"
> ```

The TPCDS q67 issue in CH is not related, please revert q67 in excluding list, thanks.
Others LGTM.

**Review by rui-mo — APPROVED** [review]

**Pull request merged**